### ISSUE TITLE
Adopt Relaton service to fetch documents

### ISSUE THREAD

**Issue by abunashir** [relaton/relaton#67]
### Background:

We're developing a new service - [Relaton Service]( https://github.com/relaton/api.relaton.org). The idea with this service is to wraps around the underlying standard's website/API, and work as a hub between those websites and our libraries. This will minimize the amount of traffic we sent to these websites and also maintain a consistent response.

So, the new relaton workflow will look like this, we will keep the local caching as it is, and if there is nothing stored locally then we will request those documents from this new service. This service maintains a temporary cache, so the response should be instant, but if not, it will communicate with the actual website, build a cache and send us the response back.

Why are we doing this?

* A centralized place / single source for all documents
* Reduce traffic to standards website, and have a consistent response.
* Use this service to build a new website and expose it as an application.

### What do we need to change in relaton?

* Use this new service to fetch document - https://github.com/relaton/api.relaton.org
* A-Parser to parse this JSON response and make it compatible with existing toolchain
* Keep the document fetching part configurable so we can switch between the actual standard library and this new API.

@andrew2net, @opoudjis, @ronaldtse: Please have a look and let me know if there is something I can help with here.

**Comment by andrew2net** [comment]
@abunashir I see the Relaton Service uses the `relaton` gem to fetch and cache documents. As I understand we going to use the `relaton` to power the Relaton Service and to fetch date form the Relaton Service, do we?
Does the API work already?

**Comment by ronaldtse** [comment]
@andrew2net the Relaton Service is an aggregator of Relaton information, i.e. the "global cache" (useful in the case of speed and also when ISO.org goes down)

1. The Relaton Service will utilize relaton-* gems to fetch relaton data to store.
2. In the future, relaton client gems can fetch from the Relaton Service as a proxy.

**Comment by abunashir** [comment]
Yes, you are right, this is a bit tricky and ideally, we should separate these in the coming days. For now, I was hoping to keep the fetching functionality configurable, so by default if will use the new API to fetch any document.

And also keep the current way of fetching a document from actual standard, so when the service itself using the relaton gem then it can enable that configuration to fetch from the actual source, and we don't create circular dependency, what do you think?

> @abunashir I see the Relaton Service uses the `relaton` gem to fetch and cache documents. As I understand we going to use the `relaton` to power the Relaton Service and to fetch date form the Relaton Service, do we?

Yes, it's using the existing relaton gem, so everything should be working as it was working locally, but the response would be in json now.
> Does the API work already?

**Comment by abunashir** [comment]
@andrew2net: Did you have a chance to look into it? Let me know if not then I might start working on this one.

**Comment by andrew2net** [comment]
@abunashir As I understand we need to implement a fetching from api.relaotn.org in the `relaton` gem, and a configuration where it should fetch documents from. So by default, the `relaton` fetches documents from api.relaton.org, but the configuration should allow switching it to the standard sites. Right?
I have a lot of tasks to do. So if you can do it let's start working on this one. If you have any questions please ask me.

**Comment by abunashir** [comment]
**TLDR**: I'm proposing a new client library, that will be simple and lightweight with caching capability. This is the only thing the user will need to install. The current `relaton` gem will only be used by the service in the future.

**Long version:**

I spend some time going through the relaton libraries, and I was hoping to integrate the new service API with the current relaton library, but after giving some thought I came across these issues:

* Adding two fetching mechanism in the current gem will add more complexity, especially the current version responds with flavor specific bibliography class, but the API service response with a generic format. So, integrating this means we have to handle both of these scenarios.

* The service is taking care of communicating with each standards, so the user shouldn't have to install each of the flavors with each installation, but if you look at this current version then it includes all of its flavors, so I don't see the point for a user having to install each of those. All they need is a lightweight and simple client that knows how to communicate with the service API and maintain the local cache.

So, here is what I would suggest:

* Create a new `relaton-client` library, this will be the entry point to the relaton toolchain.  This will mainly communicate with our service API, and respond with a Generic class similar to `Bibdata /Collection` from `relaton-cli`, this model should be interchangeable between different formats like `XML/YAML`.

* The `relaton-client` will include the local caching behavior and it will be smart enough to use the cache or do an actual request. This process should be migratable between versions and it should be decoupled from having to install each of those flavors.

* The current `relaton` gem will mainly be used by the `Relaton Service`, this can be as it is for now, but in the long run, we should remove the caching behavior from this gem, and also to come up with a more straightforward interface to fetch from a standard and having generic response.

@ronaldtse, @opoudjis, @andrew2net: thoughts?

**Comment by abunashir** [comment]
ping @ronaldtse

**Comment by ronaldtse** [comment]
@abunashir a new lightweight client library is fine, however, we have to keep the `relaton-*` gems locally because if the API Relaton site isn't available the local machine will still need to fetch, right?

**Comment by abunashir** [comment]
Cool, that's a good point, but I wouldn't add relaton flavored gems directly to the client library, and keep it simple as a medium of communication with the service API with a clean interface. We have to make sure the service has high availability with the fastest response time and on top of that the client would also keep a local cache, so I would add some analytics there and base on those take the next step to improve it.

We are still keeping the current `relaton` gem, so if that's desirable then the user can also use this library as well. In the long run, if that actually becomes important for the user then we can also add a mechanism in the CLI to choose between these libraries.
